rust/kernel/kunit.rs

```diff
@@ -88,3 +88,95 @@ macro_rules! kunit_assert_eq {
+/// can be invoked without parameters to generate the delimiter.
+#[macro_export]
+macro_rules! kunit_case {
+    () => {
+        $crate::bindings::kunit_case {
+            run_case: None,
+            name: core::ptr::null_mut(),
+            generate_params: None,
+            status: $crate::bindings::kunit_status_KUNIT_SUCCESS,
+            log: core::ptr::null_mut(),
+        }
+    };
+    ($name:ident, $run_case:ident) => {
+        $crate::bindings::kunit_case {
+            run_case: Some($run_case),
+            name: $crate::c_str!(core::stringify!($name)).as_char_ptr(),
+            generate_params: None,
+            status: $crate::bindings::kunit_status_KUNIT_SUCCESS,
+            log: core::ptr::null_mut(),
+        }
+    };
+}
+
+/// Registers a KUnit test suite.
+///
+/// # Safety
+///
+/// `test_cases` must be a NULL terminated array of test cases.
+///
+/// # Examples
+///
+/// ```ignore
+/// unsafe extern "C" fn test_fn(_test: *mut crate::bindings::kunit) {
+///     let actual = 1 + 1;
+///     let expected = 2;
+///     assert_eq!(actual, expected);
+/// }
+///
+/// static mut KUNIT_TEST_CASE: crate::bindings::kunit_case = crate::kunit_case!(name, test_fn);
+/// static mut KUNIT_NULL_CASE: crate::bindings::kunit_case = crate::kunit_case!();
+/// static mut KUNIT_TEST_CASES: &mut[crate::bindings::kunit_case] = unsafe {
+///     &mut[KUNIT_TEST_CASE, KUNIT_NULL_CASE]
+/// };
+/// crate::kunit_unsafe_test_suite!(suite_name, KUNIT_TEST_CASES);
+/// ```
+#[macro_export]
+macro_rules! kunit_unsafe_test_suite {
+    ($name:ident, $test_cases:ident) => {
+        const _: () = {
+            static KUNIT_TEST_SUITE_NAME: [i8; 256] = {
+                let name_u8 = core::stringify!($name).as_bytes();
+                let mut ret = [0; 256];
+
+                let mut i = 0;
+                while i < name_u8.len() {
+                    ret[i] = name_u8[i] as i8;
+                    i += 1;
+                }
+
+                ret
+            };
+
+            // SAFETY: `test_cases` is valid as it should be static.
+            static mut KUNIT_TEST_SUITE: $crate::bindings::kunit_suite =
+                $crate::bindings::kunit_suite {
+                    name: KUNIT_TEST_SUITE_NAME,
+                    test_cases: unsafe { $test_cases.as_mut_ptr() },
+                    suite_init: None,
+                    suite_exit: None,
+                    init: None,
+                    exit: None,
+                    status_comment: [0; 256usize],
+                    debugfs: core::ptr::null_mut(),
+                    log: core::ptr::null_mut(),
+                    suite_init_err: 0,
+                };
+
+            // SAFETY: `KUNIT_TEST_SUITE` is static.
+            #[used]
+            #[link_section = ".kunit_test_suites"]
+            static mut KUNIT_TEST_SUITE_ENTRY: *const $crate::bindings::kunit_suite =
+                unsafe { &KUNIT_TEST_SUITE };
```

Comment:
If KUnit modifies `KUNIT_TEST_SUITE`, this will need to use `&mut` and then cast to `*mut` first before casting to `*const` to avoid UB. If it doesn't get modified it could use `static` instead of `static mut`.

Reply:
If you take an immutable reference you are promising rustc that the referenced memory won't be modified until the reference ends. As such rustc would technically be allowed to demote the `static mut` to `static` (it doesn't currently to be clear, but it may in the future). By first taking a mutable reference and then casting it to a *mut raw pointer and then *const raw pointer you avoid this problem. Not that directly casting from &mut to *const will create an intermediate immutable reference somewhat confusingly. Another option is to use `addr_of_mut!(KUNIT_TEST_SUITE) as *const _` to avoid all references.

Reply:
Yeah, please use `addr_of*!` -- it is what we want to express here, anyway: a pointer, rather than a borrow.

Actually, is there a reason why `addr_of!` would not work? Then we can avoid the `*const` cast too.

Reply:
By the way:

```rust
static _A: i32 = 0;
static mut _B: *mut i32 = &_A as *const _ as *mut _;
static mut _C: *mut i32 = core::ptr::addr_of!(_A) as *mut _;
static mut _D: *mut i32 = core::ptr::addr_of_mut!(_A); // Error.
```

Is this expected to work @bjorn3? Given the current docs of `addr_of_mut!` (i.e. ignoring the implementation), I would expect so.

However, due to how the macro is implemented via `&raw`, it does not, and then the details leak into the compiler output even in the stable compiler, including mentioning raw references which would be unknown for stable users, as well as claiming a mutable borrow is attempted (but the user was using `addr_of*!` precisely to avoid those).

Reply:
`addr_of!` produces a raw pointer through which you can't write, while `addr_of_mut!` produces one through which you can write. As such `addr_of_mut!` is necessary here. It indeed seems to not be allowed in statics though. With `-Zteach` it gives an extra note:

```
error[E0764]: raw mutable references are not allowed in the final value of statics
 --> <anon>:4:38
  |
4 | static mut BAR: *const u8 = unsafe { std::ptr::addr_of_mut!(FOO) as *const u8 };
  |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: References in statics and constants may only refer to immutable values.\n\n
          Statics are shared everywhere, and if they refer to mutable data one might violate memory
          safety since holding multiple mutable references to shared data is not allowed.\n\n
          If you really want global mutable state, try using static mut or a global UnsafeCell.
  = note: this error originates in the macro `std::ptr::addr_of_mut` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error

For more information about this error, try `rustc --explain E0764`.
```

In other words using `UnsafeCell` in `KUNIT_TEST_SUITE` is probably the best way to handle this.

Reply:
`addr_of!`'s docs do not say that the produced pointer cannot be used for writing, though. I was aware of the issue of taking borrows to create pointers, but when using `addr_of!` directly to produce a `*const`, it would be fairly surprising that one is introducing extra promises. What am I missing? (i.e. how can one reach that conclusion? Or are you only talking about the current implementation of the macro?).

Reply:
From the bottom of the documentation of `addr_of!`:

> See [addr_of_mut](https://doc.rust-lang.org/stable/std/ptr/macro.addr_of_mut.html) for how to create a pointer to unininitialized data. Doing that with addr_of would not make much sense since **one could only read the data**, and that would be Undefined Behavior.

Reply:
I saw that, but that is just an example. Not only that, my reading of that is "this is a `*const` pointer, thus the type literally does not writing such as `*p = 1;`", not that nobody else could write into that place elsewhere, or that one couldn't cast it into a mutable one, initialize, and then start reading from the original pointer.

For instance, https://doc.rust-lang.org/stable/core/mem/union.MaybeUninit.html#method.as_ptr explicitly mentions writing through the returned pointer is not allowed (i.e. even if casted), but `addr_of!` does not mention that.

Reply:
That is great, thanks a ton for those.

Reply:
I was reading through those, and saw that https://github.com/rust-lang/unsafe-code-guidelines/issues/257 also discusses the `addr_of!` question (and you replied there! :)

Reply:
> (and you replied there! :)

It's hard to keep track of every discussion I ever replied on :)

Reply:
Changed to `static mut KUNIT_TEST_SUITE: core::cell::UnsafeCell<$crate::bindings::kunit_suite>`

Reply:
Since `KUNIT_TEST_SUITE_ENTRY` is a pointer you intent to write through, using `*mut` instead of `*const` would make that intent more clear.

(Just stumbled across this while following backrefs from the various issues you linked above. ;)

Also FWIW using both `static mut` and `UnsafeCell` is unnecessary -- but in this case, using `static` + `UnsafeCell` would complain about the lack of `Sync`, so you'd need a `SyncUnsafeCell` type, and using `static mut` directly you need the unstable feature `const_mut_refs` to compute the initial value for `KUNIT_TEST_SUITE_ENTRY`. So  `static mut` + `UnsafeCell`  is a useful hack right now to work around this, but once const_mut_refs is stable you should be able to remove the `UnsafeCell` and instead use `addr_of_mut!` to directly construct a `*mut`.